tools/crm_shadow.c

```diff
@@ -21,17 +21,69 @@
 #include <errno.h>
 #include <fcntl.h>
 #include <crm/msg_xml.h>
-#include <crm/common/xml.h>
 
+#include <crm/common/cmdline_internal.h>
 #include <crm/common/ipc.h>
+#include <crm/common/xml.h>
 
 #include <crm/cib.h>
 #include <crm/cib/internal.h>
 
-static int command_options = cib_sync_call;
+#define SUMMARY "perform Pacemaker configuration changes in a sandbox\n\n"  \
+                "This command sets up an environment in which "             \
+                "configuration tools (cibadmin,\n"                          \
+                "crm_resource, etc.) work offline instead of against a "    \
+                "live cluster, allowing\n"                                  \
+                "changes to be previewed and tested for side effects."
+
+#define INDENT "                              "
+
+enum shadow_command {
+    shadow_cmd_none = 0,
+    shadow_cmd_which,
+    shadow_cmd_display,
+    shadow_cmd_diff,
+    shadow_cmd_file,
+    shadow_cmd_create,
+    shadow_cmd_create_empty,
+    shadow_cmd_commit,
+    shadow_cmd_delete,
+    shadow_cmd_edit,
+    shadow_cmd_reset,
+    shadow_cmd_switch,
+};
+
+static crm_exit_t exit_code = CRM_EX_OK;
+
 static cib_t *real_cib = NULL;
-static int force_flag = 0;
-static int batch_flag = 0;
+
+static struct {
+    enum shadow_command cmd;
+    int cmd_options;
+    char *shadow;
+    bool force;
+    gboolean batch;
+    gboolean full_upload;
+    const gchar *validate_with;
+} options = {
+    .cmd_options = cib_sync_call,
+};
+
+#if 0
+// @COMPAT Enable this at next backward compatibility break
+#define SET_COMMAND(command) do {                                       \
+        if (options.cmd != shadow_cmd_none) {                           \
+            g_set_error(error, PCMK__EXITC_ERROR, CRM_EX_USAGE,         \
+                        "Only one command option may be specified");    \
+            return FALSE;                                               \
+        }                                                               \
```

Comment:
BTW I have mixed feelings about this. The two reasonable possibilities are error or last wins. Error is safer for catching user errors, but last wins lets you define a default command (via an alias or in a script) then provide overrides.

For example, I have `git recent` aliases to `log --abbrev-commit --format='%C(auto) %h %d %s' -10` (i.e. show 10 most recent commits), and since last wins for git, I can do `git recent -20` to change the number shown without an error.

Thankfully we can kick the can down the road to the next compatibility break, but we should pick an approach and stick with it then.

Reply:
(Haven't fully read your comment yet) I meant to comment here that I pulled this directly from `crm_resource.c` BTW... must not have posted it